File: codeforces/edu_r1/D.py
import sys
sys.setrecursionlimit(1000000)

pic_counter = 0
idx = 0
grid = []
vis = [[0]*1010 for x in xrange(1010)]
res = [[0]*1010 for x in xrange(1010)]
n,m = 0,0

# Total possible components
final_res = [0]*1000100

def dfs(x, y):
    global pic_counter, idx, grid, vis, res, n, m, final_res
    if(x < 0 or y < 0 or x >= n or y >= m):
        return
    if(vis[x][y]):
        return
    if(grid[x][y] == '*'):
        pic_counter += 1
        return
    vis[x][y] = 1
    dfs(x+1, y)
    dfs(x, y+1)
    dfs(x-1, y)
    dfs(x, y-1)
    res[x][y] = idx

def magic():
    global pic_counter, idx, grid, vis, res, n, m, final_res
    n,m,k = map(int, raw_input().split())
    for i in xrange(n):
        tmp = list(raw_input())
        grid.append(tmp)

    for _ in xrange(k):
        x,y = map(int,raw_input().split())
        x -= 1
        y -= 1
        if not vis[x][y]:
            idx += 1
            pic_counter = 0
            dfs(x, y)
            final_res[idx] = pic_counter
        print final_res[res[x][y]]
magic()

File: codeforces/edu_r1/D.cpp
#include <bits/stdc++.h>
using namespace std;
int n,m,k;
char grid[1010][1010];
int vis[1010][1010] = {0};
int tmp_res[1010][1010] = {0};
int cnt = 0;
int final_res[10001000] = {0};
int idx = 0; // idx for compoenent_map

void dfs(int x,int y)
{
    if(x < 0 || y < 0 || x >= n || y >= m)
    {
        return;
    }
    if(vis[x][y])
    {
        return;
    }
    if(grid[x][y] == '*')
    {
        cnt++;
        return;
    }
    vis[x][y] = 1;
    dfs(x+1, y);
    dfs(x, y+1);
    dfs(x-1, y);
    dfs(x, y-1);
    tmp_res[x][y] = idx;
}

int main()
{
    scanf(" %d %d %d", &n,&m,&k);
    for(int i=0;i<n;i++)
    {
        scanf(" %s", grid[i]);
    }
    for(int i=0;i<k;i++)
    {
        int x,y;
        scanf(" %d %d",&x,&y);
        x--;
        y--;
        if(!vis[x][y])
        {
            idx++;
            cnt = 0;
            dfs(x,y);
            final_res[idx] = cnt;
        }
        printf("%d\n",final_res[tmp_res[x][y]]);
    }
}